### PR TITLE
migration: Add case for precopy bandwidth

### PR DESCRIPTION
**Pull request by cliping**
VIRT-293955 - VM live migration - network bandwidth - precopy

Signed-off-by: cliping <lcheng@redhat.com>

### CODE REVIEW

**Comment by cliping** [comment]
Test result:

#avocado run --vt-type libvirt --test-runner=runner --vt-machine-type q35 migration.migration_efficiency.migration_precopy_bandwidth
No python imaging library installed. Screendump and Windows guest BSOD detection are disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
No python imaging library installed. PPM image conversion to JPEG disabled. In order to enable it, please install python-imaging or the equivalent for your distro.
JOB ID     : 50ff8ef9a70ce52b69c743c75f2a3c8fa410ef69
JOB LOG    : /var/lib/avocado/job-results/job-2022-09-03T09.11-50ff8ef/job.log
 (01/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.default_bandwidth.with_precopy.p2p: PASS (218.86 s)
 (02/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.default_bandwidth.with_precopy.non_p2p: PASS (218.20 s)
 (03/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_when_vm_shutoff.with_precopy.p2p: PASS (245.02 s)
 (04/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_when_vm_shutoff.with_precopy.non_p2p: PASS (243.38 s)
 (05/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_when_vm_running.with_precopy.p2p: PASS (217.08 s)
 (06/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_when_vm_running.with_precopy.non_p2p: PASS (219.77 s)
 (07/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_when_initiating_live_migration.with_precopy.p2p: PASS (215.93 s)
 (08/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_when_initiating_live_migration.with_precopy.non_p2p: PASS (220.13 s)
 (09/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_during_live_migration.with_precopy.p2p: PASS (220.05 s)
 (10/10) type_specific.io-github-autotest-libvirt.migration.migration_efficiency.migration_precopy_bandwidth.set_bandwidth_during_live_migration.with_precopy.non_p2p: PASS (217.87 s)
RESULTS    : PASS 10 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB HTML   : /var/lib/avocado/job-results/job-2022-09-03T09.11-50ff8ef/results.html
JOB TIME   : 2237.32 s

File: provider/migration/migration_base.py [hunk 1]
```diff
@@ -276,3 +277,27 @@ def check_qemu_mem_lock_hard_limit(params):
...
+def check_domjobinfo_during_mig(params):
+    """
+    Check domjobinfo during migration
+
+    :param params: dict, used to setup the connection
+    """
+    vm_name = params.get("migrate_main_vm")
+    vm = params.get("vm_obj")
+    libvirt_domjobinfo.check_domjobinfo(vm, params)
+    virsh.migrate_setspeed(vm_name, "8796093022207", debug=True)
```
Comment (Yingshun) [inline provider/migration/migration_base.py:303]: Please see my comments in https://github.com/autotest/tp-libvirt/pull/4480#discussion_r962469450.
It seems that they're total same, right??

Comment (cliping) [reply provider/migration/migration_base.py:303]: Yes, https://github.com/autotest/tp-libvirt/pull/4479 and https://github.com/autotest/tp-libvirt/pull/4480 used the same  function check_domjobinfo_during_mig().

File: libvirt/tests/src/migration/migration_efficiency/migration_precopy_bandwidth.py [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+
+        """
+        test.log.debug("Setup bandwidth when vm shutoff.")
+        migrate_desturi_port = params.get("migrate_desturi_port")
+        migrate_desturi_type = params.get("migrate_desturi_type", "tcp")
+        compare_to_value = params.get("compare_to_value")
+
+        migration_obj.conn_list.append(migration_base.setup_conn_obj(migrate_desturi_type, params, test))
+        migration_obj.remote_add_or_remove_port(migrate_desturi_port)
+        libvirt.set_vm_disk(vm, params)
```
Comment (chunfuwen) [inline libvirt/tests/src/migration/migration_efficiency/migration_precopy_bandwidth.py:29]: this will change VM config, do we need redefine VM again here?

Comment (cliping) [reply libvirt/tests/src/migration/migration_efficiency/migration_precopy_bandwidth.py:29]: VM will destroy and restart in set_vm_disk().

File: provider/migration/migration_base.py [hunk 3]
```diff
@@ -276,3 +277,27 @@ def check_qemu_mem_lock_hard_limit(params):
...
+
+def set_bandwidth_during_mig(params):
+    """
+    Set bandwidth during migration
+
+    :param params: dict, used to setup the connection
+    """
+    vm_name = params.get("migrate_main_vm")
+    compare_to_value = params.get("compare_to_value")
+    virsh.migrate_setspeed(vm_name, compare_to_value, debug=True)
```
Comment (chunfuwen) [inline provider/migration/migration_base.py:290]: Do we need check virsh.xx command status?

**Review by fangge1212 — APPROVED** [review]

File: provider/migration/migration_base.py (hunk 3 above)
Comment (cliping) [reply provider/migration/migration_base.py]: Added 'ignore_status: False'. Thanks.

**Review by chunfuwen — APPROVED** [review]
LGTM

File: provider/migration/migration_base.py (hunk 1 above)
Comment (Yingshun) [reply provider/migration/migration_base.py]: OK, they are almost same, but It seems that #4480 has more complete  check_domjobinfo_during_mig() and set_maxdowntime_during_mig(), the results will be checked in that pr. Could you please check if we should deal with 4480 first?

**Review by Yingshun — APPROVED** [review]

**Pull request merged**